### ISSUE TITLE
Windows 10 leaks DNS when using VPN by sending DNS to all network interfaces

### ISSUE THREAD

**Issue by bsclifton** [brave/brave-browser#22163]
## Description 
See https://medium.com/@ValdikSS/beware-of-windows-10-dns-resolver-and-dns-leaks-5bc5bfb4e3f1 for more information about why the "leak" happens (it's a Windows feature called `Smart Multi-Homed Name Resolution`).

Basically, Windows 10 will run multiple DNS queries (sending to multiple network interfaces) and it chooses the fastest response. Because of this behavior, 

- Brave ships with DNS over HTTPS enabled (defaulted to `With your current service provider`) which you can view on brave://settings/security
- When changing DNS over HTTPS to use `Cloudflare (1.1.1.1)`, it works as expected. 

## Possible solutions
- We could detect if VPN is connected and enable this while connected (or make a preference for that behavior). That would only affect the browser though.
- We could expose a preference (in Brave) to toggle the registry (`HKEY_LOCAL_MACHINE\Software\Policies\Microsoft\Windows NT\DNSClient`). If toggle is enabled, we can disable `Smart Multi-Homed Name Resolution` by creating a DWORD key with the name `DisableSmartNameResolution` and a value set to anything except 0. If toggle is disabled, we can remove that value

## Steps to Reproduce

   1. Be on Windows
   2. Have a VPN setup through the operating system (Windows key, type `VPN settings`, manually add one)
   3. Connect to VPN
   4. Visit https://browserleaks.com/dns and wait for test results

## Actual result:
Some DNS queries will be resolved by the ISP 🙀 

## Expected result:
DNS queries should ALL be resolved by the VPN

## Reproduces how often: 
100%

**Comment by levicki** [comment]
> We could expose a preference (in Brave) to toggle the registry (`HKEY_LOCAL_MACHINE\Software\Policies\Microsoft\Windows NT\DNSClient`). If toggle is enabled, we can disable Smart Multi-Homed Name Resolution by creating a DWORD key with the name `DisableSmartNameResolution` and a value set to anything except 0. If toggle is disabled, we can remove that value

Please do not do this.

Browser should not mess with Windows Group Policy settings for at least two reasons:

1. It will be unreliable because the PC can be part of a domain, and domain GPO can overwrite the value.
2. It can break DNS queries in complex setups and leave users wondering why something suddenly doesn't work after installing Brave.

One possible solution would be to integrate a DNS client into Brave and then make that into an option (preferrably also controllable via Group Policy) so that you can choose not to use operating system DNS client. However, note that even then the DNS queries from such a built-in client can still be blocked or redirected by the firewall.

For example, on my home network I already rewrite all direct DNS queries from all network devices to use the firewall's caching DNS server, which in turn uses CloudFlare DNS to resolve those queries.

I believe that applications should not be adding this kind of functionality, because it is hostile to enterprise environment where it reduces visibility to network administrators (especially if said applications get compromised via extensions or other means) which will just lead to all the applications that enable and/or enforce DoH being banned.

**Comment by bsclifton** [comment]
Those are great points, @levicki - thanks for the input 😄 The best solution IMO for solving is to configure DoH. 

We may try to ask users (when using VPN) to enable resolvers that aren't default in the interface and either offer a one click solution or deep link to the DoH section in settings

**Comment by levicki** [comment]
@bsclifton You are welcome.

Do note that if enabling DoH disables sending queries via Windows DNS client, then that could work for Brave.

However, that still leaves the OS DNS requests as well as a bunch of other application DNS requests, especially those based on Electron, or Microsoft Edge WebView2 which might use built-in DNS clients with their own preferred resolvers, again circumventing the VPN.

Finally, even disabling smart multi-homed resolution will only disable parallel query on all network interfaces, but if DNS response fails it will still leak the query via [LLMNR](https://www.blackhillsinfosec.com/how-to-disable-llmnr-why-you-want-to/) (and if that fails via NetBIOS) on all interfaces.

TL;DR -- privacy on the Internet is hard, doubly so if you aren't controlling the full stack from the OS to the router.

**Comment by srirambv** [comment]
@spylogsster not sure if this is fixed on 1.41.33. I still see the local IP and local DNS leak through when I check on `https://ipleak.net`. 

cc: @GeetaSarvadnya as she also repro'd the issue while testing on Windows. @brave/legacy_qa to try as well

**Comment by stephendonner** [comment]
## Verification `IN-PROGRESS` using

Brave | 1.41.91 Chromium: 103.0.5060.114 (Official Build) beta (64-bit)
-- | --
Revision | a1c2360c5b02a6d4d6ab33796ad8a268a6128226-refs/branch-heads/5060@{#1124}
OS | Windows 11 Version 21H2 (Build 22000.778)

Reproduced the original issue using `1.40.113`.

### Original issue

### Post-fix

### Confirm Cloudflare DNS servers

### `brave://settings/security` UI

### (Wireshark?)

**Comment by MadhaviSeelam** [comment]
Verification `In-progress` using
```
Brave | 1.41.91 Chromium: 103.0.5060.114 (Official Build) beta (64-bit)
-- | --
Revision | a1c2360c5b02a6d4d6ab33796ad8a268a6128226-refs/branch-heads/5060@{#1124}
OS | Windows 11 Version 21H2 (Build 22000.739)
```
Reproduced on latest release build `1.40.113`
## Steps: 
1. Install 1.41.91
2. Purchase and setup Brave VPN on `1.41.91 beta` 
3. Connect to a region - `Netherlands`)
4. launch `1.40.113`
5. load `browserleaks.com/dns` on `1.40.113`
6. confirmed "local" (ISP, i.e. non-VPN-region) DNS-server addresses shown
8. load `ipleak.net` 
9. confirm you see "local" DNS-server addresses being discovered
10. switch the VPN region on beta to another region - `UK`
11. reload `browserleaks.com/dns` 
12. reload `ipleak.net` using release
13. confirmed "local" DNS-server addresses shown

Step 3|Step 6|Step8|Step 9|Step 10|Step 13
---|---|---|---|---|---
<img width = "850" alt="Step 3" src="https://user-images.githubusercontent.com/98358127/177628757-13a2155c-786d-46c8-a1b0-ae5d99c0a138.png">|<img width = "850" alt="Step 6" src="https://user-images.githubusercontent.com/98358127/177623647-064c3310-acff-4d43-9f15-9a9fd81658d0.png">|<img width = "850" alt="Step 8" src="https://user-images.githubusercontent.com/98358127/177624073-9ecbc8ee-35db-4148-bac5-315d0a7b4e9b.png">|<img width = "850" alt="Step 9" src="https://user-images.githubusercontent.com/98358127/177625539-27cafc34-1d95-48f8-a135-69e8113c94e7.png">|<img width = "850" alt="Step 10" src="https://user-images.githubusercontent.com/98358127/177625895-e7654753-37f1-40b0-bb29-cdf487359039.png">|<img width = "850" alt="Step 13" src="https://user-images.githubusercontent.com/98358127/177625336-276972e0-db9f-4ca5-9260-9efd390a214d.png">|
